code/modules/hydroponics/compost_bin.dm

```diff
@@ -0,0 +1,195 @@
+	new /obj/item/stack/sheet/wood(loc, 10)
+	..()
+
+/obj/machinery/compost_bin/screwdriver_act(mob/living/user, obj/item/I)
+	/// there are no screws either
+	to_chat(user, "<span class='warning'>[src] has no screws!</span>")
+
+/obj/machinery/compost_bin/crowbar_act(mob/living/user, obj/item/I)
+	/// no panel either
+	return default_deconstruction_crowbar(user, I, ignore_panel = TRUE)
+
+/// takes care of plant insertion and conversion to biomass, and start composting what was inserted
+/obj/machinery/compost_bin/attackby(obj/item/O, mob/user, params)
+	if(user.a_intent == INTENT_HARM)
+		return ..()
+
+	else if(istype(O, /obj/item/storage/bag/plants))
+		if(biomass >= biomass_capacity)
+			to_chat(user, "<span class='warning'>[src] can't hold any more biomass!</span>")
+			return
+
+		var/obj/item/storage/bag/plants/PB = O
+		for(var/obj/item/reagent_containers/food/snacks/grown/G in PB.contents)
+
+			if(biomass >= biomass_capacity)
+				break
+
+			/// create biomass from plant nutriment and plant matter
+			var/plant_biomass = G.reagents.get_reagent_amount("nutriment") + G.reagents.get_reagent_amount("plantmatter")
+
+			biomass += clamp(plant_biomass*10, 1 ,biomass_capacity-biomass)
+
+			PB.remove_from_storage(G, src)
+
+		/// start composting after plants are inserted
+		compost()
+
+		if(biomass < biomass_capacity)
+			to_chat(user, "<span class='info'>You empty [PB] into [src].</span>")
+		else
+			to_chat(user, "<span class='info'>You fill [src] to its capacity.</span>")
+
+		SStgui.update_uis(src)
+		return TRUE
+
+	else if(istype(O, /obj/item/reagent_containers/food/snacks/grown))
+		if(biomass >= biomass_capacity)
+			to_chat(user, "<span class='warning'>[src] can't hold any more plants!</span>")
+			return
+		if(!user.unEquip(O))
+			return
+
+		O.forceMove(src)
+		/// make biomass from nutriment and plant matter
+		var/plant_biomass = O.reagents.get_reagent_amount("nutriment") + O.reagents.get_reagent_amount("plantmatter")
+		biomass += clamp(plant_biomass*10, 1 ,biomass_capacity-biomass)
+		qdel(O)
+		/// start composting after plants are inserted
+		compost()
+		to_chat(user, "<span class='info'>You put [O] in [src].</span>")
+		SStgui.update_uis(src)
+		return TRUE
+	else
+		to_chat(user, "<span class='warning'>You cannot put this in [name]!</span>")
+	SStgui.update_uis(src)
+	update_icon_state()
+
+
+/obj/machinery/compost_bin/ui_data(mob/user)
+	var/list/data = list(
+		"biomass" = biomass,
+		"biomass_capacity" = biomass_capacity,
+		"compost" = compost,
+		"compost_capacity" = compost_capacity,
+		)
+	return data
+
+/obj/machinery/compost_bin/attack_hand(mob/user)
+	ui_interact(user)
+
+
+/obj/machinery/compost_bin/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "CompostBin", "CompostBin", 390, 300, master_ui, state)
```

Comment:
Can you please put the ui procs in this order:
ui_interact
ui_data
ui_act

Reply:
will do